hail/python/hail/ir/base_ir.py

```diff
@@ -71,6 +85,59 @@ def _eq(self, other):
     def __hash__(self):
         return 31 + hash(str(self))
 
+    @abc.abstractmethod
+    def new_block(self, i: int) -> bool:
+        ...
+
+    @staticmethod
+    def is_effectful() -> bool:
+        return False
+
+    def binds(self, i: int) -> bool:
```

Comment:
I'm unsure what this method is supposed to do. In certain places it returns a boolean, in others a set?

Reply:
It should always return a boolean; (/agg_/scan_)bindings returns a set. I originally had this as a way to quickly check if a node binds anything, without actually constructing the dict of bindings. But I think it may not be needed anymore.

Reply:
places like here were throwing me off:
https://github.com/hail-is/hail/pull/7000/files#diff-de84e7c683cb75eeb5eaddc118d82078R75

> I originally had this as a way to quickly check if a node binds anything,

I had guessed this, but wasn't sure! Thanks.

Reply:
Ah, right. Currently it's only used in `BaseIR.child_context`, which adds new bindings to the current context. It allows a short-cut in the common case where nothing is added. But I think that's not worth the complexity. I'll remove it.

Reply:
Oh, wow, it appears I was somebody else while editing blockmatrix_ir. I'm looking forward to getting back to a statically typed language.